src/main/java/net/mcreator/ui/dialogs/StateEditorDialog.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * MCreator (https://mcreator.net/)
+ * Copyright (C) 2012-2020, Pylo
+ * Copyright (C) 2020-2021, Pylo, opensource contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.ui.dialogs;
+
+import net.mcreator.ui.MCreator;
+import net.mcreator.ui.component.util.PanelUtils;
+import net.mcreator.ui.help.HelpUtils;
+import net.mcreator.ui.help.IHelpContext;
+import net.mcreator.ui.init.L10N;
+import net.mcreator.ui.minecraft.states.PropertyData;
+
+import javax.swing.*;
+import java.awt.*;
+import java.util.List;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+public class StateEditorDialog {
+
+	public static String open(MCreator parent, String initialState, Map<String, PropertyData> properties, String help) {
+		final String stateString = initialState.equals("") ? "!esc" : initialState;

```

Comment:
Will you use `!esc` and `!new` in the other PRs? I don't see where it's used. Also maybe instead of writing directly those default behaviour, you can add some constants instead and use them.

Reply:
What I meant by constants is something like `public static final String ESC = "!esc";`.

Reply:
Also, storing state? in string like this is a really smelly code

Reply:
How else should I store it? I remember AI condition editor returns ArrayList with 2 Strings for procedure names, why is it bad way here?
>What I meant by constants is something like `public static final String ESC = "!esc";`.

This could be a way to go, but my code also checks for tokens in general (if the result starts with ! - a token string).

Reply:
I understood. This could be passed directly into retVal and then its getter will be used instead of that string.